### PR TITLE
Shebang

### PR DESCRIPTION
**Pull request by kzidane**
@dmalan any takes?

Inspired by https://github.com/vim/vim/blob/071d4279d6ab81b7187b48f3a0fc61e587b6db6c/runtime/scripts.vim#L31

### CODE REVIEW

**Comment by kzidane** [comment]
Actually, there's a bug when opening multiple files at once. Working on a fix...

File: simple.js [hunk 1]
```diff
@@ -496,6 +498,48 @@ define(function(require, exports, module) {
...
+         * Detects Ace syntax mode from shebang for files with no extension
+         */
+        function detectSyntax() {
+            // syntax modes
+            var MODES = {
+                perl: /perl/,
+                php: /php/,
+                python: /python/,
+                ruby: /ruby/,
+                sh: /^(bash2?|ksh|sh)/
```
Comment (dmalan) [inline simple.js]: `csh`, `tcsh`

File: simple.js [hunk 2]
```diff
@@ -496,6 +498,48 @@ define(function(require, exports, module) {
...
+                if (tab.editorType !== "ace" || !_.isEmpty(extname(tab.path)))
+                    return;
+
+                var editor = tab.editor;
+                var session = editor.ace.getSession();
+
+                // ensure first line is a shebang
+                var firstLine = session.getLine(0);
+                if (firstLine.startsWith("#!")) {
+                    // turn lines like #!/usr/bin/env [VAR=val]... bash to #!/usr/bin/bash for easy matching
```
Comment (dmalan) [inline simple.js]: This seems presumptuous if the binary doesn't actually live at that path. Does it not suffice to check for a format of

`#!/path/to/foo [bar ...]`

and then pattern match against `foo` if `bar` is `undefined`, else if `foo` is `env` and `bar` is not `undefined`, then pattern-match against `bar`?

File: simple.js (hunk 1 above)
Comment (kzidane) [inline simple.js]: Apparently vim [doesn't think](https://github.com/vim/vim/blob/071d4279d6ab81b7187b48f3a0fc61e587b6db6c/runtime/scripts.vim#L58-L67) they're the same? Can add anyway.

Comment (dmalan) [inline simple.js]: Oh, not the same, not sure if C9 can syntax-highlight properly, just mentioning them as other shells. `zsh` too.

Comment (kzidane) [inline simple.js]: Well, C9's syntax mode is named [`sh`](https://github.com/c9/core/blob/master/node_modules/ace/lib/ace/mode/sh.js). So I only took `sh` along with everything else that vim groups with it.

**Comment by kzidane** [comment]
Turns out `ace` already [detects](https://github.com/c9/core/blob/106832d7b6f5f10fd83553db34e84c6478b42af3/plugins/c9.ide.ace/ace.js#L1434) some syntax modes from shebang. It doesn't worry too much about format though (not sure if there's a reason).

Proposed adding more modes and suggested improving. See c9/core#374.

Closing this for now!

**Pull request closed without merging**